range.go

```diff
@@ -0,0 +1,88 @@
+package tiledb
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+// Range is an 1D range along a subarray dimension
+type Range struct {
```

Comment:
Would it make sense to have the range type be generic? That means the user can’t accidentally unpack the wrong type of range.

Reply:
I would love to but go does not allow generic methods, cannot do `(sa *Subarray) methods[T](v T)`

Reply:
You define methods on the generic type, so it would look like:

```go
func (sa *Subarray[T]) DoSomething() (T, error) {
  // ...
}
```

See [a complete example I banged together](https://go.dev/play/p/EFXirsf_3Yh) in the Go Playground.

Reply:
Correct. A subarray can have dimensions of different types. Should we go for types `Subarray1[T], Subarray2[T, V]`?

Reply:
`AddRange` is smooth and can be implemented with full static type checking. `GetRange` introduces an asymmetry, it cannot return by definition anything typed so it is up to the caller to use the proper `ExtractRange`. We can handle such issues if we rewrite `Array` with generics and keep track of the dimension types. This can be a followup.

The other thing is `Var` handling. True `Range` and `Range with var` are different. However if we implement them this way we will have types `Range, RangeVar` and methods `AddRange, AddRangeVar, GetRange, GetRangeVar` Now the methods are uniform, and we have one type of `Range` which we use in helpers.

Reply:
I now see the problem that you’re describing, since (as you note) the difficult part is getting a useful Range from the Subarray itself. I had not fully considered the implications of the roundtrip.